Fix #11750 (refactor ctu-info to generate fewer artifacts on disk)

danmar opened the pull request.

firewave commented:
Please add some Python tests which for those files after an analysis.

I wanted to add these to test some local changes but those are already included in this PR.

firewave commented on cli/cppcheckexecutor.cpp:
```diff
@@ -170,9 +174,14 @@ namespace {
+        /**
+         * CTU information
+         */
+        std::string mCtuInfo;
```
Is this safe in terms of memory usage? I have not really understood what this is doing yet (maybe add some explanation to the PR) but appears to accumulate the the data in the memory and this could be megabytes in size or even much, much more.

danmar replied:
we should check it but my hypothesis is that the ctu info will not be huge.

danmar replied:
about memory usage, here is one small test case: cppcheck/test/cli/whole-program
 * the files whole1.c and whole2.c are 94 and 64 bytes.
 * the ctu-info that is generated for those are 220 bytes and 227 bytes.

in this test case ~160 bytes source code generates ~450 bytes ctu-info

I don't think that memory usage will be a large issue. If we pretend that scanning a large project with 160MB source code would require 450MB memory for ctu info.

firewave replied:
Thanks.

But that has to be constructed as a string and is required to be continuous memory.

And we also need to profile that. That sounds like it might slow down things quite a bit. Maybe we might need both modes?

danmar commented:
I 100% agree about adding more testing. I will do it. But I don't even manage to run our original tests yet, those do test this functionality also.

firewave commented:
> I 100% agree about adding more testing. I will do it. But I don't even manage to run our original tests yet, those do test this functionality also.

Possible - as I mentioned I have not looked into yet.

danmar commented on test/cli/whole-program_test.py:
```diff
@@ -214,6 +214,21 @@ def test_suppress_inline_project_builddir_j(tmpdir):
+    _, _, stderr = cppcheck(args, cwd=__script_dir)
+    assert 'misra-c2012-5.8' in stderr
+    _, _, stderr = cppcheck(args, cwd=__script_dir)
+    assert 'misra-c2012-5.8' in stderr
```
My intention was only to refactor. But here is a test case that does not work in cppcheck main branch but does work in this branch.

danmar commented:
there is testing for CTU analysis in addons in test/cli/whole-program_test.py

the tests in that file broke when I started working on this refactoring.

firewave commented:
> there is testing for CTU analysis in addons in test/cli/whole-program_test.py
> 
> the tests in that file broke when I started working on this refactoring.

Yes - I added those. There are also far from complete and have known issues which need to be fixed. So "breaking" them might actually be fixing those (I might not have not written all tests with XFAIL).

I was referring to tests which check what exists on the disk and is left (or not) on it and not just the analysis results.

The local changes I have is getting rid of the duplicated cleanup code and some further reduction of redundancies.

danmar commented:
> I was referring to tests which check what exists on the disk and is left (or not) on it and not just the analysis results.

no cppcheck build dir: I don't see the point to write a test that tests no files are saved on the disk. the code and tests should be connected. I feel it would be like adding a test that checks that files.txt is not created in various directories.

cppcheck build dir: I add one more test

danmar commented:
> no cppcheck build dir: I don't see the point to write a test that tests no files are saved on the disk. the code and tests should be connected. I feel it would be like adding a test that checks that files.txt is not created in various directories.

I added a test that no artifacts is remaining in the project folder. And locally that test fails. So I need to reconsider this..

firewave commented on gui/threadhandler.h:
```diff
@@ -257,6 +257,7 @@ protected slots:
     int mRunningThreadCount{};
 
     bool mAnalyseWholeProgram{};
+    QString mCtuInfo;
```
It looks like this is never set and thus always empty.

danmar replied:
good catch :+1: I need to investigate this

firewave commented on gui/checkthread.cpp:
```diff
@@ -117,10 +117,11 @@ void CheckThread::check(const Settings &settings)
 {
     mFiles = files;
     mAnalyseWholeProgram = true;
+    mCtuInfo = ctuInfo;
```
Since this is always converted to a `std::string` on usage I think it makes sense to store at as such if we are copying it anyways.

danmar replied:
done.

firewave commented on test/cli/whole-program_test.py:
```diff
@@ -214,6 +216,65 @@ def test_suppress_inline_project_builddir_j(tmpdir):
+def test_addon_rerun(tmpdir, jobs, builddir):
+    """Rerun analysis and ensure that misra CTU works; with and without build dir, with -j1 and -j2"""
+    args = [
+        f'-j{jobs}',
```
We are not using format strings.

danmar replied:
alright :-(

firewave replied:
Some context: they were just introduced in Python 3.6 and we still support Python 3.5 (via Ubuntu 16.04) although we currently cannot test that in the CI.

firewave replied:
We currently do not have documented what Python versions are supported. Since support for Python 2.7 is removed in this development cycle I will add that information and then we can decide if we want to deprecate more afterwards.

danmar replied:
thanks.. for user scripts i.e. the addons I I vote to support as old python version as we can. For these test scripts I don't feel it's that important. If we can execute these test scripts in the CI with whatever python version is readily available there that is fine for me.

firewave replied:
> the addons I I vote to support as old python version as we can.

There is an upcoming PR which will clarify things. I want to finish up with the pre-3.x Python stuff first.

firewave commented on test/cli/whole-program_test.py:
```diff
@@ -214,6 +216,65 @@ def test_suppress_inline_project_builddir_j(tmpdir):
+    _, _, stderr = cppcheck(args, cwd=__script_dir)
+    assert 'misra-c2012-5.8' in stderr
+
+@pytest.mark.parametrize("jobs", (1,2))
```
No need to explicitly test `-j`. That is done automatically within the CI.

danmar replied:
you mean that the CI will add -j options in the args I create in the test case. I would not think so.

firewave replied:
That was added in #6061.

And as pointed out somewhere else this should probably be done for `--cppcheck-build-dir` as well as that should not change the behavior in single run execution. But that is for later.

danmar replied:
you miss the point as far as I see. I want to test with a single thread and I ALSO want to test with multiple threads. So I need to remove that TEST_INJECT environment variable if that is set.

danmar replied:
As far as I see I don't need to remove TEST_CPPCHECK_INJECT_J because it has this code:

    if 'TEST_CPPCHECK_INJECT_J' in os.environ:
        found_j = False
        for arg in args:
            if arg.startswith('-j'):
                found_j = True
                break
        if not found_j:
            arg_j = '-j' + str(os.environ['TEST_CPPCHECK_INJECT_J'])
            args.append(arg_j)

so TEST_CPPCHECK_INJECT_J should have no effect if explict -j option is used as far as I see.

danmar replied:
:+1: I have updated

firewave commented:
I only reviewed the general stuff. I still need to dig into the build dir stuff. Also still needs to be performance tested.

danmar commented on test/cli/whole-program_test.py:
```diff
@@ -214,6 +216,65 @@ def test_suppress_inline_project_builddir_j(tmpdir):
+@pytest.mark.parametrize("jobs,builddir", ((1,False), (1,True), (2,False), (2,True)))
+def test_addon_no_artifacts(tmpdir, jobs, builddir):
+    """Test that there are no artifacts left after analysis"""
+    shutil.copyfile(os.path.join(__script_dir, 'whole-program', 'whole1.c'), os.path.join(tmpdir, 'whole1.c'))
```
because there can be artifacts in the test folder before I start the test. Want to have a clean folder with just the test files. This test is also executed locally and locally you might have artifacts..

firewave replied:
That seems unexpected.

#6787 is about detecting such leftovers.

danmar replied:
But #6787 only works in the CI. what if I add a file locally one way or another in the test folder. For instance by running cppcheck and pressing Ctrl+C while there is some temporary dump file or something.
That is not tested in the CI.

firewave replied:
Yes, but it is a start (weirdly it doesn't fail). But so we at least know that it works as expected in the default (i.e. no fatal errors/interruption) case. We can move forward from that.

danmar replied:
> Yes, but it is a start (weirdly it doesn't fail).

sure. I don't disapprove that PR of course.

danmar commented on lib/cppcheck.cpp:
```diff
@@ -1426,11 +1425,13 @@ void CppCheck::executeAddons(const std::vector<std::string>& files, const std::s
             continue;
 
-        const std::vector<picojson::value> results =
+        const std::vector<picojson::value>& results =
```
hmm no I thought why do we want to copy by value. is there a good reason? I reverted this change.

danmar commented on test/cli/whole-program_test.py:
```diff
@@ -214,6 +216,65 @@ def test_suppress_inline_project_builddir_j(tmpdir):
+    _, _, stderr = cppcheck(args, cwd=__script_dir)
+    assert 'misra-c2012-5.8' not in stderr
+
+@pytest.mark.parametrize("jobs,builddir", ((1,False), (1,True), (2,False), (2,True)))
```
:+1: I have updated

danmar commented:
I didn't expect that performance would be affected much. But according to my measurements this makes cppcheck faster at least in a self-check.

I built cppcheck from this branch and main branch using the Makefile using this build command:
```
make CXXFLAGS=-O2 MATCHCOMPILER=yes
```

And I saw faster analysis with the cppcheck binary built from this branch (cppcheck-11750) than from main branch (cppcheck-HEAD):
```
cppcheck-11750 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    6m37,058s
user    6m17,025s
sys     0m20,182s
cppcheck-11750 -j2 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    3m55,146s
user    6m17,953s
sys     0m24,346s
cppcheck-11750 -j8 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    1m57,996s
user    8m5,076s
sys     0m27,413s
cppcheck-HEAD --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    7m14,262s
user    6m54,306s
sys     0m20,013s
cppcheck-HEAD -j2 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    4m28,635s
user    6m45,882s
sys     0m23,421s
cppcheck-HEAD -j8 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    2m29,994s
user    8m11,637s
sys     0m25,597s
```

danmar commented on test/cli/whole-program_test.py:
```diff
@@ -214,6 +216,65 @@ def test_suppress_inline_project_builddir_j(tmpdir):
     os.mkdir(build_dir)
     __test_suppress_inline_project(tmpdir, ['-j2', '--cppcheck-build-dir={}'.format(build_dir)])
 
+@pytest.mark.parametrize("jobs,builddir", ((1,False), (1,True), (2,False), (2,True)))
```
:+1: I have updated

firewave commented:
> And I saw faster analysis with the cppcheck binary built from this branch (cppcheck-11750) than from main branch (cppcheck-HEAD)

Thanks for doing those tests.

But you cannot analyze the `lib` from the repo because you modified that - so you are comparing apples and slightly-not-apples. You need a fixed corpus for both runs.

Beside that I would not expect such an improvement from the code changes so that seems extremely suspect and I assume there is some data being omitted from analysis. Best would be to store the `--debug` output from a `-j1` run and compare that.

danmar commented:
> But you cannot analyze the lib from the repo because you modified that - so you are comparing apples and slightly-not-apples. You need a fixed corpus for both runs.

I compiled cppcheck-11750 and cppcheck-HEAD first and then I ran it all in 1 go from a script:
```
$ cat run.sh

echo "cppcheck-11750 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib"
time ./cppcheck-11750 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib -q 2> /dev/null
echo "cppcheck-11750 -j2 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib"
time ./cppcheck-11750 -j2 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib -q 2> /dev/null
echo "cppcheck-11750 -j8 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib"
time ./cppcheck-11750 -j8 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib -q 2> /dev/null

echo "cppcheck-HEAD --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib"
time ./cppcheck-HEAD --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib -q 2> /dev/null
echo "cppcheck-HEAD -j2 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib"
time ./cppcheck-HEAD -j2 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib -q 2> /dev/null
echo "cppcheck-HEAD -j8 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib"
time ./cppcheck-HEAD -j8 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib -q 2> /dev/null
```
So it should be the same files. I ate lunch while it was running so I wasn't modifying files in the meantime.

danmar commented:
> Best would be to store the --debug output from a -j1 run and compare that.

hmm how would --debug output be useful does that contain any relevant timing info?

> Beside that I would not expect such an improvement from the code changes so that seems extremely suspect

Frankly I didn't expect it neither. I didn't expect much difference in performance at all.

danmar commented:
> Frankly I didn't expect it neither. I didn't expect much difference in performance at all.

oh wait I will have to execute different addons.

firewave commented:
> hmm how would --debug output be useful does that contain any relevant timing info?

Since that shows the data we analyze - but that would only be actual code and not what we pass as CTU - so that is of no use. Maybe we should add a debug option which displays which CTU information we generate and pass to the analysis.

danmar commented:
I have updated the script.

```
cppcheck-11750 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib1

real    5m11,995s
user    4m52,916s
sys     0m19,250s
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
cppcheck-main --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib1

real    5m22,830s
user    5m3,381s
sys     0m19,665s
DONE
```

The output is:

```
cppcheck-11750 --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib1

real    5m10,728s
user    4m53,270s
sys     0m17,675s
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
cppcheck-main --premium=misra-c++-2023 -D__GNUC__ -D__CPPCHECK__ lib

real    5m29,503s
user    5m11,975s
sys     0m17,692s
DONE
```

The speedup compared to the previous run can be explained because I ran release builds of premiumaddon this time.

firewave commented:
> The speedup compared to the previous run can be explained because I ran release builds of premiumaddon this time.

That result makes way more sense. Thanks for clearing that up.

firewave commented on test/cli/whole-program_test.py:
```diff
@@ -210,6 +212,61 @@ def test_suppress_inline_project_builddir_j(tmpdir):
     __test_suppress_inline_project(tmpdir, ['-j2', '--cppcheck-build-dir={}'.format(build_dir)])
 
+@pytest.mark.parametrize("builddir", (False,True))
+def test_addon_rerun(tmpdir, builddir):
```
If we are adding new Python tests it would make sense to use the modern `tmp_path` instead.

The main difference is that you use '/' to concat paths and you might need to explicitly convert it to `str` in some cases. Otherwise not much is different. See 0a36cdd493e50c4d33182457f55e51f0f087e523 for some tests I added recently.

danmar replied:
done

firewave commented:
I still would like to have a final look but I am out for most of the next two days.

danmar commented:
@firewave friendly ping

danmar commented:
@firewave friendly ping

firewave commented:
My concerns have been address but  I still cannot wrap my head around the actual workflow. My mind just isn't really there. I will give it another spin tomorrow by debugging through it.

firewave commented on cli/cppcheckexecutor.cpp:
```diff
@@ -129,6 +129,10 @@ namespace {
         }
 
+        const std::string& getCtuInfo() const {
+            return mCtuInfo;
```
It feels a bit strange that this lives in the just remotely related logger. I wonder if all the logic should live in its own object and cleanup should be RAII-based and such.

That would be similar to what I tried in #6634 which comes up short in making it a local object because it needs to intercept the logged error at some point (which also applies to the code here). So adding some kind of hook into the error logging which std, XML and CTU (and even SARIF) could use would make a lot of sense.

danmar replied:
I am not sure if I understand these comments fully.

I do feel the ctu info belongs in cli and gui rather than lib. it's aggregated info from all threads. For me it's reasonable that CppcheckExecutor owns the data (directly or indirectly) however I can agree that the functionality does not technically belong in StdLogger actually.

We also have the mActiveCheckers and mCriticalErrors. Those do not really fit very well in the StdLogger neither.

I do not see a very elegant way to insert hooks. We could add one more ErrorLogger that owns the StdLogger?

     class AllLogger : public ErrorLogger {
     private:
         void reportErr(const ErrorMessage& errmsg) {
              ... handle active checkers, critical errors, ctu, ...

              // pass remaining errors to stdLogger
              stdLogger.reportErr(errmsg);
         }
         StdLogger stdLogger;
    };

Do you have better ideas..?

danmar replied:
Personally I still suggest to put the condition in StdLogger that ensures the output is formatted with sarif/text/xml so it will take a ErrorMessage input and output the proper string..

firewave replied:
Yeah, again something in this file turns into a grab-bag after things got sorted out at some point.

This is not something we need to solve in this PR. I will see if I am put together a PR with the proposed hook based on the previously linked cleanup.

firewave commented:
The changes seem fine to me. That behavior is not changed much and it seems the test coverage is sufficient so if somewhere were wrong that should show up.

I will add the build dir injection for all tests soon and also test the before and after. So if there are corner case that should be (hopefully) uncovered by this.

firewave commented:
I forgot the essential in my last comment: Feel free to merge. 👍🙂

danmar commented:
Thanks!

firewave commented on lib/cppcheck.cpp:
```diff
@@ -1418,8 +1418,7 @@ void CppCheck::executeAddons(const std::vector<std::string>& files, const std::s
     std::string fileList;
 
     if (files.size() >= 2 || endsWith(files[0], ".ctu-info")) {
-        fileList = Path::getPathFromFilename(files[0]) + FILELIST + std::to_string(mSettings.pid);
```
Removing the unique identifier (pid) from the filename re-introduces the issue this was fixing - parallel executions writing the same file. This is exposed by running the Python tests with `python-xdist` - see #6696.

danmar replied:
oh no!! :-(

andymacg replied:
I think I am seeing the same bug, but not totally sure.

This repo helps reproduce a concurrency problem on my Linux machine
https://github.com/andymacg/cppcheck-addon-concurrency-bug

firewave replied:
A quick fix is simply adding back the pid to the filename. Will take a look later - I totally forgot about it.

danmar replied:
> A quick fix is simply adding back the pid to the filename. Will take a look later - I totally forgot about it.

I have done that with #7082. Feel free to review it..
